### PR TITLE
New Resource: Add DDos Protection Plan for Virtual Network.

### PR DESCRIPTION
**Pull request by metacpp**
This PR introduces the new resource (requested in #2584, `azurerm_ddos_protection_plan`, which is only allowed per region.

There will be another PR to associate and enable DDos Protection Plan for Virtual Network.

### CODE REVIEW

**Comment by metacpp** [comment]
> Does you description indicate there can be only one?
Yes, and I grouped all the tests into group test.

File: azurerm/resource_arm_virtual_network.go [hunk 1]
```diff
@@ -183,6 +203,8 @@ func resourceArmVirtualNetworkRead(d *schema.ResourceData, meta interface{}) err
 			d.Set("address_space", space.AddressPrefixes)
 		}
 
+		d.Set("ddos_protection_plan", flattenVirtualNetworkDDosProtectionPlan(props))
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_network.go]: As this is a complex object could we check err here?
```suggestion
		if err := d.Set("ddos_protection_plan", flattenVirtualNetworkDDosProtectionPlan(props)); err != nil {
			return fmt.Errorf("Error setting `ddos_protection_plan`: %+v", err)
		}
```

File: azurerm/resource_arm_virtual_network.go [hunk 2]
```diff
@@ -48,6 +48,26 @@ func resourceArmVirtualNetwork() *schema.Resource {
...
+			"ddos_protection_plan": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_network.go]: Could we validate this as an id: `azure.ValidateResourceID`

File: azurerm/resource_arm_virtual_network.go [hunk 3]
```diff
@@ -48,6 +48,26 @@ func resourceArmVirtualNetwork() *schema.Resource {
...
 
+			"ddos_protection_plan": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:         schema.TypeString,
+							Optional:     true,
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_network.go]: Would this be better as `required`? to clear the value we just wouldn't set this block?

**Comment by hbuckle** [comment]
@katbyte, @metacpp I was developing something else against a build of master and I think I found a small  bug in this - Azure seems to default ddos protection to false even when not specified, so you always get a diff

```hcl
resource "azurerm_resource_group" "resource_group" {
  name     = "hbtest_vnet"
  location = "uksouth"
}

resource "azurerm_virtual_network" "virtual_network" {
  name                = "hbtest_vnet"
  address_space       = ["10.0.0.0/16"]
  location            = "${azurerm_resource_group.resource_group.location}"
  resource_group_name = "${azurerm_resource_group.resource_group.name}"
}
```

```cmd
terraform apply -auto-approve

azurerm_resource_group.resource_group: Creating...
  location: "" => "uksouth"
  name:     "" => "hbtest_vnet"
  tags.%:   "" => "<computed>"
azurerm_resource_group.resource_group: Creation complete after 1s (ID: /subscriptions/020ef497-e33d-4705-859b-08663256a4bc/resourceGroups/hbtest_vnet)
azurerm_virtual_network.virtual_network: Creating...
  address_space.#:     "" => "1"
  address_space.0:     "" => "10.0.0.0/16"
  location:            "" => "uksouth"
  name:                "" => "hbtest_vnet"
  resource_group_name: "" => "hbtest_vnet"
  subnet.#:            "" => "<computed>"
  tags.%:              "" => "<computed>"
azurerm_virtual_network.virtual_network: Still creating... (10s elapsed)
azurerm_virtual_network.virtual_network: Creation complete after 11s (ID: /subscriptions/020ef497-e33d-4705-859b-...ft.Network/virtualNetworks/hbtest_vnet)

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

terraform plan

Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

azurerm_resource_group.resource_group: Refreshing state... (ID: /subscriptions/020ef497-e33d-4705-859b-08663256a4bc/resourceGroups/hbtest_vnet)
azurerm_virtual_network.virtual_network: Refreshing state... (ID: /subscriptions/020ef497-e33d-4705-859b-...ft.Network/virtualNetworks/hbtest_vnet)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ azurerm_virtual_network.virtual_network
      ddos_protection_plan.#: "1" => "0"


Plan: 0 to add, 1 to change, 0 to destroy.
```

Json returned from the API

```json
{
  "name": "hbtest_vnet",
  "id": "/subscriptions/xxx/resourceGroups/hbtest_vnet/providers/Microsoft.Network/virtualNetworks/hbtest_vnet",
  "etag": "W/\"0611dcfd-bfc6-43cd-a2af-434ab17f827e\"",
  "type": "Microsoft.Network/virtualNetworks",
  "location": "uksouth",
  "tags": {},
  "properties": {
    "provisioningState": "Succeeded",
    "resourceGuid": "98c5a26b-ab3d-49f9-8564-e15b424f3d37",
    "addressSpace": {
      "addressPrefixes": [
        "10.0.0.0/16"
      ]
    },
    "dhcpOptions": {
      "dnsServers": []
    },
    "virtualNetworkPeerings": [],
    "enableDdosProtection": false,
    "enableVmProtection": false
  }
}
```

**Comment by metacpp** [comment]
@hbuckle thanks for reporting this issue, I can reproduce it and I've prepared a quick fix for it.

File: azurerm/provider.go [hunk 4]
```diff
@@ -196,6 +196,7 @@ func Provider() terraform.ResourceProvider {
 			"azurerm_data_lake_store_firewall_rule":          resourceArmDataLakeStoreFirewallRule(),
 			"azurerm_data_lake_store":                        resourceArmDataLakeStore(),
 			"azurerm_databricks_workspace":                   resourceArmDatabricksWorkspace(),
+			"azurerm_ddos_protection_plan":                   resourceArmDDosProtectoinPlan(),
```
Comment (rakelkar) [inline azurerm/provider.go]: >resourceArmDDosProtectoinPlan [](start = 53, length = 29)

typo :)

**Pull request merged**